docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+		// Record a new sender for a tx we've already seen.
+		// Note it's possible a tx is still in the cache but no longer in the mempool
+		// (eg. after committing a block, txs are removed from mempool but not cache),
+		// so we only record the sender for txs still in the mempool.
+		if e, ok := mem.txsMap.Load(tx.Key()); ok {
+			memTx := e.(*clist.CElement).Value.(*mempoolTx)
+			memTx.senders.LoadOrStore(txInfo.SenderID, true)
+			// TODO: consider punishing peer for dups,
+			// its non-trivial since invalid txs can become valid,
+			// but they can spam the same tx with little cost to them atm.
+		}
+
+    // If transaction was invalid, we need to remember the peer information
+    if _, ok := mem.cache.invalidCachedTx.Load(tx.Key); ok {
+        mem.banPeer(peerID)  
+    }
+		return mempool.ErrTxInCache
+	}
+
+```
+
+
+```golang
+
+    func (mem* ClistMempool) banPeer(peerID NodeID) {
+        numFails := 0
+         // Check whether this peer has sent us transactions that fail
+        if val, ok := mem.peerFailureMap[peerID]; ok {
+         lastFailureT := val.lastFailure
+         numFails = val.numFails
+        // if the failure was recent enough, update the number of failures and 
+        // ban peer if applicable
+        if time.Since(lastFailureT) <= failureResetInterval {
+          if numFails == maxAllowedFailures - 1 {
+              // Send Ban request to p2p 
+            }
+          }
+        }
+        // Update the time of the last failure     
+        mem.peerFailureMap[peerID] = { time.Now(), numFailures + 1}
+    }
+```
+If transactions with `ResponseCheckTx.code > 1` are not deleted from the cache and we want to ban them on the first offence,
+we can skip the second `if` in the code above but call `banPeer` immediately at the end of the function. The function `banPeer` should simply forward the `peerID` to the p2p layer. In fact, we do not need to store any information on this peer as the node will remove it from its peer set.
+
+**Signaling the ban to p2p**
+
+As it is the mempool **reactor** that has access to the p2p layer, not the actual mempool implementation, the peer banning function will most likely have to send the peer ID to a channel to inform the reactor of the fact that this peer should be banned. This would require adding a channel for communication into the `CListMempool` on construction, and adding a routine in the mempool ractor that waits on a response via this channel. However, the actual implementation of this is yet to be defined.
+
+**Implementing peer banning on recheck**
+
+Currently the recheck logic confirmes whether once **valid** transactions, 
+, where `ResponseCheckTx.code == 0`, are still valid. 
+
+As this logic loops through the transactions in any case, we can leverage it to check whether we can ban peers. 
+However, this approach has several downsides:
+- It is not timely enough. Recheck is executed after a block is committed, leaving room for a bad peer to send 
```

Comment:
As mentioned above, my thought on this part is that I can't to see a valid example of an App that _locally_ passed `CheckTx` on a TX, and then fails at `reCheckTx` time with "> 1"

Reply:
No, we would not expect the transaction to fail it here after it passed it the first time. What I meant here is that we check whether the transaction was marked as invalid with a particular code when it was added, and if yes we ban all the peers that sent it. So we only leverage the fact that we anyways loop through all the transactions and do not apply the normal recheck logic (which applies to only valid transactions).

In my opinion this solution is a bad one, but I decided to list it here in case people feel differently.